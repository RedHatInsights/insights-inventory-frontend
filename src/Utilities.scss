// Sass Imports

@import 'Utilities/_variables.scss';
@import 'Utilities/_colors.scss';
@import 'Utilities/_helpers.scss';
@import 'Utilities/_mixins.scss';

@import '~@redhat-cloud-services/frontend-components/index.css';
