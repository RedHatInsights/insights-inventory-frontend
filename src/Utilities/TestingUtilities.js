import React from 'react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';

export const mountWithRouter = (Component, initialEntries) => {
  const wrapper = mount(
    <MemoryRouter initialEntriest={initialEntries}>{Component}</MemoryRouter>
  );

  return wrapper;
};

export const TestWrapper = ({
  children,
  routerProps = { initialEntries: ['/'] },
  path,
  store = {},
}) => {
  return (
    <MemoryRouter {...routerProps}>
      <Provider store={store}>
        {path ? (
          <Routes>
            <Route path={path} element={children} />
          </Routes>
        ) : (
          children
        )}
      </Provider>
    </MemoryRouter>
  );
};

TestWrapper.propTypes = {
  children: PropTypes.any.isRequired,
  routerProps: PropTypes.shape({
    initialEntries: PropTypes.arrayOf(PropTypes.string),
  }),
  path: PropTypes.string,
  store: PropTypes.any,
};
