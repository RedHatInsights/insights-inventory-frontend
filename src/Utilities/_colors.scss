// Sample Colors
$light: var(--pf-global--BackgroundColor--300);
$medium: #393f44;
$dark: #292e34;

$primary: #39a5dc;
$primary-dark: #0088ce;
$primary-shadow: #00659c;

$accent: #2A2E57;
$accent-dark: #20253b;
