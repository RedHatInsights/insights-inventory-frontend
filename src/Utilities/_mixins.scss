@use 'sass:meta';
@use 'sass:math';
@use 'sass:list';
@use 'sass:string';

// Convert px to rem
// =================================================================/
@mixin rem($property, $values) {

  // Create a couple of empty lists as output buffers.
  $font-size: 14px;
  $px-values: ();
  $rem-values: ();

  // Loop through the $values list
  @each $value in $values {

    // For each property value, if it's in rem or px, derive both rem and
    // px values for it and add those to the end of the appropriate buffer.
    // Ensure all pixel values are rounded to the nearest pixel.
    @if meta.type-of($value) == number and not math.is-unitless($value) and (math.unit($value) == px) {

      // px value given - calculate rem value from font-size
      $new-rem-value: calc($value / $font-size);
      $px-values: list.join($px-values, round($value));
      $rem-values: list.join($rem-values, string.unquote("#{$new-rem-value}rem"));

    } @else if meta.type-of($value) == number and not math.is-unitless($value) and (math.unit($value) == '%') {

      // % value given - don't add px or rem
      $px-values: list.join($px-values, string.unquote(#{$value}));
      $rem-values: list.join($rem-values, string.unquote(#{$value}));

    } @else if $value == auto {

      // auto - don't add px or rem
      $px-values: list.join($px-values, auto);
      $rem-values: list.join($rem-values, auto);

    } @else {

      // math.is-unitless value - use those directly as rem and calculate the px-fallback
      $px-values: list.join($px-values, round($value * $font-size));
      $rem-values: list.join($rem-values, string.unquote("#{$value}rem"));
    }
  }

  // output the converted rules
  #{$property}: $px-values;
  #{$property}: $rem-values;

}



// Generate font-size in rem
// =================================================================/
@mixin font-size($size, $line: $size * 1.4) {
  font-size: ($size);
  font-size: ($size / $global-font-size) + rem;

  @if $line == 1 {
    line-height: 1;

  } @else if $line != null {
    line-height: $line;
    line-height: ($line / $global-font-size) + rem;
  }
}



// Generate transition
// =================================================================/
@mixin transition($property) {

  transition: $property 0.2s cubic-bezier(0.4, 0, 0.2, 1);
  -o-transition: $property 0.2s cubic-bezier(0.4, 0, 0.2, 1);
  -ms-transition: $property 0.2s cubic-bezier(0.4, 0, 0.2, 1);
  -moz-transition: $property 0.2s cubic-bezier(0.4, 0, 0.2, 1);
  -webkit-transition: $property 0.2s cubic-bezier(0.4, 0, 0.2, 1);
}