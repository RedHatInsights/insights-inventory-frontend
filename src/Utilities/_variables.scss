// Global Variables
$global-padding: 30px;
$global-margin: 15px;

$global-font-size: 14px;

// PF Font
$pf-v5-global--enable-font-overpass-cdn: true;