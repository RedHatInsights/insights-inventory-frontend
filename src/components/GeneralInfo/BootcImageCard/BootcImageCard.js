import React from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import LoadingCard from '../LoadingCard';
import { bootcSelector } from '../selectors';
import { extraShape } from '../../../constants';

export const BootcImageCard = ({ handleClick, extra }) => {
  const { detailLoaded, bootc } = useSelector(
    ({ systemProfileStore: { systemProfile } }) => ({
      detailLoaded: systemProfile && systemProfile.loaded,
      bootc: bootcSelector(systemProfile),
    })
  );

  return (
    <LoadingCard
      title="BOOTC"
      isLoading={!detailLoaded}
      items={[
        ...[{ title: 'Booted Image', value: bootc.bootedImage }],
        ...[{ title: 'Booted Image Digest', value: bootc.bootedImageDigest }],
        ...[{ title: 'Staged Image', value: bootc.stagedImage }],
        ...[{ title: 'Staged Image Digest', value: bootc.stagedImageDigest }],
        ...[{ title: 'Available Image', value: bootc.availableImage }],
        ...[
          {
            title: 'Available Image Digest',
            value: bootc.availableImageDigest,
          },
        ],
        ...[{ title: 'Rollback Image', value: bootc.rollbackImage }],
        ...[
          { title: 'Rollback Image Digest', value: bootc.rollbackImageDigest },
        ],
        ...extra.map(({ onClick, ...item }) => ({
          ...item,
          ...(onClick && { onClick: (e) => onClick(e, handleClick) }),
        })),
      ]}
    />
  );
};

BootcImageCard.propTypes = {
  detailLoaded: PropTypes.bool,
  handleClick: PropTypes.func,
  bootc: PropTypes.shape({
    bootedImage: PropTypes.string,
    bootedImageDigest: PropTypes.string,
    stagedImage: PropTypes.string,
    stagedImageDigest: PropTypes.string,
    availableImage: PropTypes.string,
    availableImageDigest: PropTypes.string,
    rollbackImage: PropTypes.string,
    rollbackImageDigest: PropTypes.string,
  }),
  extra: PropTypes.arrayOf(extraShape),
};
BootcImageCard.defaultProps = {
  detailLoaded: false,
  handleClick: () => undefined,
  extra: [],
};
