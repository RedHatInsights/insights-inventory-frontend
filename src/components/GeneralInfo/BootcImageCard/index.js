export * from './BootcImageCard';
export { default } from './BootcImageCard';
