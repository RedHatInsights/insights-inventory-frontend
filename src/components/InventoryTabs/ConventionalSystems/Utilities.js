import flatMap from 'lodash/flatMap';
import {
  HOST_GROUP_CHIP,
  RHCD_FILTER_KEY,
  UPDATE_METHOD_KEY,
} from '../../../Utilities/constants';

const mapTags = ({ category, values }) =>
  values.map(
    ({ tagKey, value }) =>
      `${category ? `${category}/` : ''}${tagKey}${value ? `=${value}` : ''}`
  );

const filterMapper = {
  staleFilter: ({ staleFilter }, searchParams) =>
    staleFilter.forEach((item) => searchParams.append('status', item)),
  osFilter: ({ osFilter }, searchParams) => {
    // TODO This is very hackish. There should be a schema/feature on how the values translate into params
    const osParams = Object.entries(
      Object.values(osFilter).reduce((acc, item) => ({ ...acc, ...item }), {})
    )
      .filter(([, value]) => value === true)
      .map(([key]) => key);

    osParams.forEach((item) => {
      const keyParts = item.split('-');
      const [osName, value] = [
        keyParts.slice(0, keyParts.length - 2).join(' '),
        keyParts[keyParts.length - 1],
      ];

      if (value && osName) {
        searchParams.append('operating_system', `${osName}${value}`);
      }
    });
  },
  registeredWithFilter: ({ registeredWithFilter }, searchParams) =>
    registeredWithFilter?.forEach((item) =>
      searchParams.append('source', item)
    ),
  value: ({ value, filter }, searchParams) =>
    value === 'hostname_or_id' &&
    Boolean(filter) &&
    searchParams.append('hostname_or_id', filter),
  tagFilters: ({ tagFilters }, searchParams) =>
    tagFilters?.length > 0 &&
    searchParams.append('tags', flatMap(tagFilters, mapTags)),
  rhcdFilter: ({ rhcdFilter }, searchParams) =>
    rhcdFilter?.forEach((item) => searchParams.append(RHCD_FILTER_KEY, item)),
  lastSeenFilter: ({ lastSeenFilter }, searchParams) =>
    Object.keys(lastSeenFilter || {})?.forEach(
      (item) =>
        item === 'mark' &&
        searchParams.append('last_seen', lastSeenFilter[item])
    ),
  updateMethodFilter: ({ updateMethodFilter }, searchParams) =>
    updateMethodFilter?.forEach((item) =>
      searchParams.append(UPDATE_METHOD_KEY, item)
    ),
  hostGroupFilter: ({ hostGroupFilter }, searchParams) =>
    hostGroupFilter?.forEach((item) =>
      searchParams.append(HOST_GROUP_CHIP, item)
    ),
};

// TODO It might be better to not use a SearchParam instance and mutate it
// This should rather just return a plain object that can be
// applied with react routers useSearch params.
export const calculateFilters = (searchParams, filters = []) => {
  filters.forEach((filter) => {
    Object.keys(filter).forEach((key) => {
      filterMapper?.[key]?.(filter, searchParams);
    });
  });
  return searchParams;
};

export const calculatePagination = (searchParams, page, perPage) => {
  const currSearch = new URLSearchParams(location.search);
  const newPage = page !== undefined ? page : currSearch.get('page');
  const newPerPage =
    perPage !== undefined ? perPage : currSearch.get('per_page');
  !isNaN(parseInt(newPage)) && searchParams.append('page', newPage);
  !isNaN(parseInt(newPerPage)) && searchParams.append('per_page', newPerPage);
};
