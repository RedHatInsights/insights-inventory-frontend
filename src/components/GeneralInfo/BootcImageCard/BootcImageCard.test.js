import { render } from '@testing-library/react';
import React from 'react';
import configureStore from 'redux-mock-store';
import { TestWrapper } from '../../../Utilities/TestingUtilities';
import { BootcImageCard } from './BootcImageCard';

describe('BootcImageCard', () => {
  let initialState;
  let mockStore;

  beforeEach(() => {
    mockStore = configureStore();
    initialState = {
      systemProfileStore: {
        systemProfile: {
          loaded: true,
          bootc_status: {
            booted: {
              image: 'quay.io:5000/bootc-insights:latest',
              image_digest:
                'sha256:8fa5b818d1560c7d15d8744069651b671acd13ec5290c2bb55d1fae2492fcb5f',
            },
            rollback: {
              image: 'quay.io:5000/bootc-insights:latest',
              cached_image: 'quay.io:5000/bootc-insights:latest',
              image_digest:
                'sha256:99dec597bd1e95565d9df181cab9bf0278b15e79a613dbd0a357d60f295e9a72',
              cached_image_digest:
                'sha256:06462b5728c3b445df63327451c32874d946eb1f2071401680145d52e578137b',
            },
          },
          cpu_flags: ['one'],
        },
      },
    };
  });

  it('should render correctly - no data', () => {
    const store = mockStore({ systemProfileStore: {} });
    const view = render(
      <TestWrapper store={store}>
        <BootcImageCard />
      </TestWrapper>
    );
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should render correctly with data', () => {
    const store = mockStore(initialState);
    const view = render(
      <TestWrapper store={store}>
        <BootcImageCard />
      </TestWrapper>
    );
    expect(view.asFragment()).toMatchSnapshot();
  });
});
