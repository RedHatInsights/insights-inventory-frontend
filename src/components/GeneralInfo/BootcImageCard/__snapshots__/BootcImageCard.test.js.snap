// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BootcImageCard should render correctly - no data 1`] = `
<DocumentFragment>
  <div
    class="pf-v5-c-card"
    data-ouia-component-id="system-properties-card"
    data-ouia-component-type="PF5/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-v5-c-card__body"
    >
      <div
        class="pf-v5-l-stack pf-m-gutter"
      >
        <div
          class="pf-v5-l-stack__item"
        >
          <div
            class="pf-v5-c-content"
          >
            <h1
              class=""
              data-ouia-component-id="SystemPropertiesCardTitle"
              data-ouia-component-type="PF5/Text"
              data-ouia-safe="true"
              data-pf-content="true"
            >
              BOOTC
            </h1>
          </div>
        </div>
        <div
          class="pf-v5-l-stack__item pf-m-fill"
        >
          <div
            class="pf-v5-c-content"
          >
            <dl
              class=""
            >
              <dt
                aria-label="Booted Image title"
                class=""
                data-ouia-component-id="Booted Image title"
              >
                Booted Image
              </dt>
              <dd
                aria-label="Booted Image value"
                class=""
                data-ouia-component-id="Booted Image value"
              >
                <div
                  class="pf-v5-c-skeleton ins-c-skeleton ins-c-skeleton__sm"
                >
                  <span
                    class="pf-v5-screen-reader"
                  />
                </div>
              </dd>
              <dt
                aria-label="Booted Image Digest title"
                class=""
                data-ouia-component-id="Booted Image Digest title"
              >
                Booted Image Digest
              </dt>
              <dd
                aria-label="Booted Image Digest value"
                class=""
                data-ouia-component-id="Booted Image Digest value"
              >
                <div
                  class="pf-v5-c-skeleton ins-c-skeleton ins-c-skeleton__sm"
                >
                  <span
                    class="pf-v5-screen-reader"
                  />
                </div>
              </dd>
              <dt
                aria-label="Staged Image title"
                class=""
                data-ouia-component-id="Staged Image title"
              >
                Staged Image
              </dt>
              <dd
                aria-label="Staged Image value"
                class=""
                data-ouia-component-id="Staged Image value"
              >
                <div
                  class="pf-v5-c-skeleton ins-c-skeleton ins-c-skeleton__sm"
                >
                  <span
                    class="pf-v5-screen-reader"
                  />
                </div>
              </dd>
              <dt
                aria-label="Staged Image Digest title"
                class=""
                data-ouia-component-id="Staged Image Digest title"
              >
                Staged Image Digest
              </dt>
              <dd
                aria-label="Staged Image Digest value"
                class=""
                data-ouia-component-id="Staged Image Digest value"
              >
                <div
                  class="pf-v5-c-skeleton ins-c-skeleton ins-c-skeleton__sm"
                >
                  <span
                    class="pf-v5-screen-reader"
                  />
                </div>
              </dd>
              <dt
                aria-label="Available Image title"
                class=""
                data-ouia-component-id="Available Image title"
              >
                Available Image
              </dt>
              <dd
                aria-label="Available Image value"
                class=""
                data-ouia-component-id="Available Image value"
              >
                <div
                  class="pf-v5-c-skeleton ins-c-skeleton ins-c-skeleton__sm"
                >
                  <span
                    class="pf-v5-screen-reader"
                  />
                </div>
              </dd>
              <dt
                aria-label="Available Image Digest title"
                class=""
                data-ouia-component-id="Available Image Digest title"
              >
                Available Image Digest
              </dt>
              <dd
                aria-label="Available Image Digest value"
                class=""
                data-ouia-component-id="Available Image Digest value"
              >
                <div
                  class="pf-v5-c-skeleton ins-c-skeleton ins-c-skeleton__sm"
                >
                  <span
                    class="pf-v5-screen-reader"
                  />
                </div>
              </dd>
              <dt
                aria-label="Rollback Image title"
                class=""
                data-ouia-component-id="Rollback Image title"
              >
                Rollback Image
              </dt>
              <dd
                aria-label="Rollback Image value"
                class=""
                data-ouia-component-id="Rollback Image value"
              >
                <div
                  class="pf-v5-c-skeleton ins-c-skeleton ins-c-skeleton__sm"
                >
                  <span
                    class="pf-v5-screen-reader"
                  />
                </div>
              </dd>
              <dt
                aria-label="Rollback Image Digest title"
                class=""
                data-ouia-component-id="Rollback Image Digest title"
              >
                Rollback Image Digest
              </dt>
              <dd
                aria-label="Rollback Image Digest value"
                class=""
                data-ouia-component-id="Rollback Image Digest value"
              >
                <div
                  class="pf-v5-c-skeleton ins-c-skeleton ins-c-skeleton__sm"
                >
                  <span
                    class="pf-v5-screen-reader"
                  />
                </div>
              </dd>
            </dl>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BootcImageCard should render correctly with data 1`] = `
<DocumentFragment>
  <div
    class="pf-v5-c-card"
    data-ouia-component-id="system-properties-card"
    data-ouia-component-type="PF5/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-v5-c-card__body"
    >
      <div
        class="pf-v5-l-stack pf-m-gutter"
      >
        <div
          class="pf-v5-l-stack__item"
        >
          <div
            class="pf-v5-c-content"
          >
            <h1
              class=""
              data-ouia-component-id="SystemPropertiesCardTitle"
              data-ouia-component-type="PF5/Text"
              data-ouia-safe="true"
              data-pf-content="true"
            >
              BOOTC
            </h1>
          </div>
        </div>
        <div
          class="pf-v5-l-stack__item pf-m-fill"
        >
          <div
            class="pf-v5-c-content"
          >
            <dl
              class=""
            >
              <dt
                aria-label="Booted Image title"
                class=""
                data-ouia-component-id="Booted Image title"
              >
                Booted Image
              </dt>
              <dd
                aria-label="Booted Image value"
                class=""
                data-ouia-component-id="Booted Image value"
              >
                quay.io:5000/bootc-insights:latest
              </dd>
              <dt
                aria-label="Booted Image Digest title"
                class=""
                data-ouia-component-id="Booted Image Digest title"
              >
                Booted Image Digest
              </dt>
              <dd
                aria-label="Booted Image Digest value"
                class=""
                data-ouia-component-id="Booted Image Digest value"
              >
                sha256:8fa5b818d1560c7d15d8744069651b671acd13ec5290c2bb55d1fae2492fcb5f
              </dd>
              <dt
                aria-label="Staged Image title"
                class=""
                data-ouia-component-id="Staged Image title"
              >
                Staged Image
              </dt>
              <dd
                aria-label="Staged Image value"
                class=""
                data-ouia-component-id="Staged Image value"
              >
                Not available
              </dd>
              <dt
                aria-label="Staged Image Digest title"
                class=""
                data-ouia-component-id="Staged Image Digest title"
              >
                Staged Image Digest
              </dt>
              <dd
                aria-label="Staged Image Digest value"
                class=""
                data-ouia-component-id="Staged Image Digest value"
              >
                Not available
              </dd>
              <dt
                aria-label="Available Image title"
                class=""
                data-ouia-component-id="Available Image title"
              >
                Available Image
              </dt>
              <dd
                aria-label="Available Image value"
                class=""
                data-ouia-component-id="Available Image value"
              >
                Not available
              </dd>
              <dt
                aria-label="Available Image Digest title"
                class=""
                data-ouia-component-id="Available Image Digest title"
              >
                Available Image Digest
              </dt>
              <dd
                aria-label="Available Image Digest value"
                class=""
                data-ouia-component-id="Available Image Digest value"
              >
                Not available
              </dd>
              <dt
                aria-label="Rollback Image title"
                class=""
                data-ouia-component-id="Rollback Image title"
              >
                Rollback Image
              </dt>
              <dd
                aria-label="Rollback Image value"
                class=""
                data-ouia-component-id="Rollback Image value"
              >
                quay.io:5000/bootc-insights:latest
              </dd>
              <dt
                aria-label="Rollback Image Digest title"
                class=""
                data-ouia-component-id="Rollback Image Digest title"
              >
                Rollback Image Digest
              </dt>
              <dd
                aria-label="Rollback Image Digest value"
                class=""
                data-ouia-component-id="Rollback Image Digest value"
              >
                sha256:99dec597bd1e95565d9df181cab9bf0278b15e79a613dbd0a357d60f295e9a72
              </dd>
            </dl>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
